Lightweight variant of:
commit 3957565a968c5eacb611940a2146fd4491d584c1
Author: Jan Kratochvil <jan.kratochvil@redhat.com>
Date:   Fri Oct 14 07:58:58 2011 +0000

    gdb/
        Drop lazy lm_info reading.

It fixes gdb.base/coredump-filter.exp on i686 and x86_64-m32.

--- gdb-7.2-orig/gdb/solib-svr4.c	2015-10-04 20:49:59.336753982 +0200
+++ gdb-7.2/gdb/solib-svr4.c	2015-10-04 22:12:38.246028656 +0200
@@ -1179,6 +1179,7 @@ svr4_current_sos (void)
       struct so_list *new = XZALLOC (struct so_list);
       struct cleanup *old_chain = make_cleanup (xfree, new);
       CORE_ADDR next_lm;
+      int err;
 
       new->lm_info = xmalloc (sizeof (struct lm_info));
       make_cleanup (xfree, new->lm_info);
@@ -1188,11 +1189,18 @@ svr4_current_sos (void)
       new->lm_info->lm = xzalloc (lmo->link_map_size);
       make_cleanup (xfree, new->lm_info->lm);
 
-      read_memory (lm, new->lm_info->lm, lmo->link_map_size);
+      err = target_read_memory (lm, new->lm_info->lm, lmo->link_map_size);
 
       next_lm = LM_NEXT (new);
 
-      if (LM_PREV (new) != prev_lm)
+      if (err != 0)
+	{
+	  warning (_("Error reading shared library list entry at %s"),
+		   paddress (target_gdbarch, lm)),
+	  free_so (new);
+	  next_lm = 0;
+	}
+      else if (LM_PREV (new) != prev_lm)
 	{
 	  warning (_("Corrupted shared library list"));
 	  free_so (new);
